Future incompatibility

Running Rust 1.62.1's cargo check reports a future incompatibility, with the following report:
```
The package `win-crypto-ng v0.4.0` currently triggers the following future incompatibility lints:
> warning: reference to packed field is unaligned
>   --> /home/nora/.cargo/registry/src/github.com-1ecc6299db9ec823/win-crypto-ng-0.4.0/src/helpers/blob.rs:38:18
>    |
> 38 |         unsafe { &self.0 }
>    |                  ^^^^^^^
>    |
>    = note: `#[allow(unaligned_references)]` on by default
>    = warning: this was previously accepted by the compiler but is being phased out; it will become a hard error in a future release!
>    = note: for more information, see issue #82523 <https://github.com/rust-lang/rust/issues/82523>
>    = note: fields of packed structs are not properly aligned, and creating a misaligned reference is undefined behavior (even if that reference is never dereferenced)
>    = help: copy the field contents to a local variable, or replace the reference with a raw pointer and use `read_unaligned`/`write_unaligned` (loads and stores via `*p` must be properly aligned even when using raw pointers)
> 
```
I don't know when this will become a problem, but I think you'd like to know :)

I think, we now just hit the next level: I am having a job running on _v1.64_ and the code does no longer compile:
```bash
error[E0793]: reference to packed field is unaligned
  --> C:\GitLab-Runner\builds\biomedit\sett-rs\.cargo\registry\src\index.crates.io-6f17d22bba15001f\win-crypto-ng-0.4.0\src\helpers\blob.rs:38:18
   |
38 |         unsafe { &self.0 }
   |                  ^^^^^^^
   |
   = note: fields of packed structs are not properly aligned, and creating a misaligned reference is undefined behavior (even if that reference is never dereferenced)
   = help: copy the field contents to a local variable, or replace the reference with a raw pointer and use `read_unaligned`/`write_unaligned` (loads and stores via `*p` must be properly aligned even when using raw pointers)
For more information about this error, try `rustc --explain E0793`.
error: could not compile `win-crypto-ng` due to previous error
```
@emgre any chance to get this fixed soon? I mean, obviously there is even a `MR` ready which should solve the issue. What are the alternatives here?

--- a/.github/workflows/ci.yml
+++ b/.github/workflows/ci.yml
@@ -6,7 +6,7 @@ jobs:
     strategy:
       matrix:
         rust:
-          - 1.37.0  # MSRV
+          - 1.60.0  # MSRV
           - stable
           - beta
           - nightly
@@ -15,9 +15,6 @@ jobs:
           - "zeroize"
           - "rand"
           - "block-cipher"
-        exclude: # Zeroize 1.1 supports Rust 1.39+
-          - rust: 1.37.0
-            features: "zeroize"
     steps:
     - uses: actions/checkout@v1
     - uses: actions-rs/toolchain@v1

--- a/Cargo.toml
+++ b/Cargo.toml
@@ -2,7 +2,7 @@
 name = "win-crypto-ng"
 version = "0.4.0"
 authors = ["Émile Grégoire <eg@emilegregoire.ca>", "Igor Matuszewski <Xanewok@gmail.com>"]
-edition = "2018"
+edition = "2021"
 description = "Safe bindings to Windows Cryptography API: Next Generation"
 repository = "https://github.com/emgre/win-crypto-ng"
 documentation = "https://docs.rs/crate/win-crypto-ng"
@@ -10,6 +10,7 @@ readme = "README.md"
 keywords = ["cng", "windows", "cryptoapi"]
 categories = ["api-bindings", "os::windows-apis", "cryptography"]
 license = "BSD-3-Clause"
+rust-version = "1.60"
 
 [badges]
 maintenance = { status = "experimental" }

--- a/README.md
+++ b/README.md
@@ -2,7 +2,7 @@
 
 [![crates.io](https://img.shields.io/crates/v/win-crypto-ng.svg)](https://crates.io/crates/win-crypto-ng)
 [![docs.rs](https://docs.rs/win-crypto-ng/badge.svg)](https://docs.rs/crate/win-crypto-ng)
-![MSRV](https://img.shields.io/badge/rustc-1.37+-blue.svg)
+![MSRV](https://img.shields.io/badge/rustc-1.60+-blue.svg)
 [![Build status](https://github.com/emgre/win-crypto-ng/workflows/CI/badge.svg)](https://github.com/emgre/win-crypto-ng/actions)
 [![License](https://img.shields.io/github/license/emgre/win-crypto-ng)](https://github.com/emgre/win-crypto-ng/blob/master/LICENSE.md)
 

--- a/src/helpers/blob.rs
+++ b/src/helpers/blob.rs
@@ -71,9 +71,7 @@ unsafe impl<T: BlobLayout> FromBytes for Blob<T> {
     // Require that the allocation is at least as aligned as its header to
     // safely reference it as the first field. (despite
     // `Blob` being technically `#[repr(packed)]`)
-    unsafe fn min_layout() -> Layout {
-        Layout::new::<T::Header>()
-    }
+    const MIN_LAYOUT: Layout = Layout::new::<T::Header>();
 
     unsafe fn ptr_cast(source: *const [u8]) -> *const Self {
         assert_ne!(source as *const (), ptr::null());

--- a/src/helpers/blob.rs
+++ b/src/helpers/blob.rs
@@ -4,6 +4,7 @@ use crate::helpers::{AsBytes, FromBytes};
 
 use core::alloc::Layout;
 use core::{mem, ptr, slice};
+use std::ptr::addr_of;
 
 /// C-compatible dynamic inline structure.
 ///
@@ -33,9 +34,11 @@ impl<'a, T: BlobLayout> Blob<T> {
         // `Self::from_boxed`, which requires that the source reference is
         // aligned at least as `T::Header` and since `Blob` is
         // `#[repr(C)]` (and so is `T::Header`, because it  implements `AsBytes`
-        // which requires being `#[repr(C)]`), so the reference to its first
-        // field will be aligned at least as `T::Header`.
-        unsafe { &self.0 }
+        // which requires being `#[repr(C)]`), so the pointer to its first
+        // field will be aligned at least as `T::Header`, and can be safely
+        // cast to a reference
+        let header_ptr = addr_of!(self.0);
+        unsafe { &*header_ptr }
     }
 
     pub(crate) fn tail(&self) -> &[u8] {

--- a/src/helpers/blob.rs
+++ b/src/helpers/blob.rs
@@ -134,11 +134,11 @@ macro_rules! blob {
                 let mut boxed = vec![0u8; header_len + tail_len].into_boxed_slice();
 
                 let header_bytes = $crate::helpers::AsBytes::as_bytes(header);
-                &mut boxed[..header_len].copy_from_slice(header_bytes);
+                boxed[..header_len].copy_from_slice(header_bytes);
                 let mut offset = header_len;
                 $(
                     let field_len = blob! { size: header, $($len)*};
-                    &mut boxed[offset..offset + field_len].copy_from_slice(tail.$field);
+                    boxed[offset..offset + field_len].copy_from_slice(tail.$field);
                     offset += field_len;
                 )*
 

--- a/src/helpers/blob.rs
+++ b/src/helpers/blob.rs
@@ -49,9 +49,6 @@ impl<'a, T: BlobLayout> Blob<T> {
         AsBytes::as_bytes(self)
     }
 
-    // False positive for arbitrary self type
-    // TODO: Remove once we bump MSRV to a newer clippy
-    #[allow(clippy::wrong_self_convention)]
     pub fn into_bytes(self: Box<Self>) -> Box<[u8]> {
         AsBytes::into_bytes(self)
     }

--- a/src/helpers/bytes.rs
+++ b/src/helpers/bytes.rs
@@ -68,8 +68,8 @@ pub unsafe trait AsBytes {
 /// Marker trait for types that can be safely converted to bytes.
 pub unsafe trait FromBytes {
     /// Specified the minimum layout requirements for the allocation:
-    /// - is at least as big as `min_layout().size()`
-    /// - referenced/pointed data is at least as aligned as `min_layout().align()`
+    /// - is at least as big as `MIN_LAYOUT.size()`
+    /// - referenced/pointed data is at least as aligned as `MIN_LAYOUT.align()`
     ///
     /// For DSTs, final size should be exactly the same as the allocation's.
     ///
@@ -77,15 +77,14 @@ pub unsafe trait FromBytes {
     /// In Rust, it is considered [UB] for pointers/references/`Box<T>`es to
     /// be dangling, unaligned or pointing to invalid value.
     ///
-    /// The alignment check is done using `min_layout` and thus can cause UB if
+    /// The alignment check is done using `MIN_LAYOUT` and thus can cause UB if
     /// implemented incorrectly.
     ///
     /// [UB]: https://doc.rust-lang.org/reference/behavior-considered-undefined.html
-    // FIXME: Replace with a const once we bump MSRV
-    unsafe fn min_layout() -> Layout;
+    const MIN_LAYOUT: Layout;
 
     fn from_bytes(bytes: &[u8]) -> &Self {
-        let min_layout = unsafe { Self::min_layout() };
+        let min_layout = Self::MIN_LAYOUT;
         // Make sure the allocation meets the expected layout requirements
         assert!(bytes.len() >= min_layout.size());
         assert_eq!(bytes.as_ptr() as usize % min_layout.align(), 0);
@@ -104,7 +103,7 @@ pub unsafe trait FromBytes {
     }
 
     fn from_boxed(boxed: Box<[u8]>) -> Box<Self> {
-        let min_layout = unsafe { Self::min_layout() };
+        let min_layout = Self::MIN_LAYOUT;
         // Make sure the allocation meets the expected layout requirements
         assert!(boxed.len() >= min_layout.size());
         assert_eq!(boxed.as_ptr() as usize % min_layout.align(), 0);
@@ -129,20 +128,18 @@ pub unsafe trait FromBytes {
 }
 
 unsafe impl FromBytes for [u16] {
-    unsafe fn min_layout() -> Layout {
-        // Allow for empty slices but require correct alignment
-        Layout::from_size_align_unchecked(0, mem::align_of::<u16>())
-    }
+    // Allow for empty slices but require correct alignment
+    const MIN_LAYOUT: Layout =
+        unsafe { Layout::from_size_align_unchecked(0, mem::align_of::<u16>()) };
     unsafe fn ptr_cast(source: *const [u8]) -> *const Self {
         ptr_slice_cast(source)
     }
 }
 
 unsafe impl FromBytes for [u8] {
-    unsafe fn min_layout() -> Layout {
-        // Allow for empty slices but require correct alignment
-        Layout::from_size_align_unchecked(0, mem::align_of::<u8>())
-    }
+    // Allow for empty slices but require correct alignment
+    const MIN_LAYOUT: Layout =
+        unsafe { Layout::from_size_align_unchecked(0, mem::align_of::<u8>()) };
     unsafe fn ptr_cast(source: *const [u8]) -> *const [u8] {
         source
     }
@@ -167,9 +164,7 @@ unsafe impl<T> FromBytes for T
 where
     T: Pod,
 {
-    unsafe fn min_layout() -> Layout {
-        Layout::new::<Self>()
-    }
+    const MIN_LAYOUT: Layout = Layout::new::<Self>();
 
     unsafe fn ptr_cast(ptr: *const [u8]) -> *const Self {
         ptr_ref_cast(ptr as *const ())

--- a/src/helpers/bytes.rs
+++ b/src/helpers/bytes.rs
@@ -87,7 +87,7 @@ pub unsafe trait FromBytes {
     fn from_bytes(bytes: &[u8]) -> &Self {
         let min_layout = unsafe { Self::min_layout() };
         // Make sure the allocation meets the expected layout requirements
-        assert!(bytes.len() >= min_layout.size(), 0);
+        assert!(bytes.len() >= min_layout.size());
         assert_eq!(bytes.as_ptr() as usize % min_layout.align(), 0);
 
         let old_size = mem::size_of_val(bytes);
@@ -106,7 +106,7 @@ pub unsafe trait FromBytes {
     fn from_boxed(boxed: Box<[u8]>) -> Box<Self> {
         let min_layout = unsafe { Self::min_layout() };
         // Make sure the allocation meets the expected layout requirements
-        assert!(boxed.len() >= min_layout.size(), 0);
+        assert!(boxed.len() >= min_layout.size());
         assert_eq!(boxed.as_ptr() as usize % min_layout.align(), 0);
 
         let old_size = mem::size_of_val(boxed.as_ref());

